core/lib/src/rocket.rs

```diff
@@ -792,10 +781,42 @@ impl Rocket {
             .executor(runtime.executor())
             .serve(service);
 
-        // TODO.async: Use with_graceful_shutdown, and let launch() return a Result<(), Error>
-        runtime.block_on(server).expect("TODO.async handle error");
+        let (future, handle) = server.remote_handle();
+        runtime.spawn(future);
+        Ok(handle.await?)
+    }
 
-        unreachable!("the call to `block_on` should block on success")
+    /// Starts the application server and begins listening for and dispatching
+    /// requests to mounted routes and catchers. Unless there is an error, this
+    /// function does not return and blocks until program termination.
+    ///
+    /// # Error
+    ///
+    /// If there is a problem starting the application, a [`LaunchError`] is
+    /// returned. Note that a value of type `LaunchError` panics if dropped
+    /// without first being inspected. See the [`LaunchError`] documentation for
+    /// more information.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # if false {
+    /// rocket::ignite().launch();
+    /// # }
+    /// ```
+    pub fn launch(self) -> LaunchError {
+        // TODO.async What meaning should config.workers have now?
+        // Initialize the tokio runtime
+        let runtime = tokio::runtime::Builder::new()
+            .core_threads(self.config.workers as usize)
+            .build()
+            .expect("Cannot build runtime!");
+
+        // TODO.async: Use with_graceful_shutdown, and let launch() return a Result<(), Error>
+        match runtime.block_on(self.launch_on(&runtime)) {
+            Ok(_) => unreachable!(),
```

Comment:
This should explain why: `unreachable!("a listening server should never complete")`

Reply:
Whoops, thought I copied the text over from before the commits. There used to be a message there, I'll use that.